modular/manager/bucket_migrate_scheduler.go

```diff
@@ -290,77 +287,93 @@ func (s *BucketMigrateScheduler) Start() error {
 	return nil
 }
 
-func (s *BucketMigrateScheduler) subscribeEvents() {
-	subscribeBucketMigrateEventsTicker := time.NewTicker(time.Duration(s.manager.subscribeBucketMigrateEventInterval) * time.Millisecond)
-	for {
-		<-subscribeBucketMigrateEventsTicker.C
-		var (
-			migrationBucketEvents []*types.ListMigrateBucketEvents
-			err                   error
-			executePlan           *BucketMigrateExecutePlan
-		)
-
-		// 1. subscribe migrate bucket events
-		migrationBucketEvents, err = s.manager.baseApp.GfSpClient().ListMigrateBucketEvents(context.Background(), s.lastSubscribedBlockHeight+1, s.selfSP.GetId())
+func (s *BucketMigrateScheduler) processEvents(migrateBucketEvents *types.ListMigrateBucketEvents) error {
+	// 1. process CancelEvents
+	if migrateBucketEvents.CancelEvents != nil {
+		executePlan, err := s.getExecutePlanByBucketID(migrateBucketEvents.CancelEvents.BucketId.Uint64())
 		if err != nil {
-			log.Errorw("failed to list migrate bucket events", "error", err)
-			continue
+			log.Errorw("bucket migrate schedule received EventCompleteMigrationBucket", "CancelEvents", migrateBucketEvents.CancelEvents)
+			return err
+		}
+		s.deleteExecutePlanByBucketID(migrateBucketEvents.CancelEvents.BucketId.Uint64())
+		executePlan.stopSPSchedule()
+	}
+	// 2. process CompleteEvents
+	if migrateBucketEvents.CompleteEvents != nil {
+		executePlan, err := s.getExecutePlanByBucketID(migrateBucketEvents.CompleteEvents.BucketId.Uint64())
+		// 1) Received the CompleteEvents event for the first time.
+		// 2) Subsequently received the CompleteEvents event.
+		if err != nil {
+			log.Errorw("bucket migrate schedule received EventCompleteMigrationBucket, the event may already finished", "CompleteEvents", migrateBucketEvents.CompleteEvents)
+			return err
 		}
 
-		// 2. make plan, start plan
-		for _, migrateBucketEvents := range migrationBucketEvents {
-			// when receive chain CompleteMigrationBucket event
-			if migrateBucketEvents.CompleteEvents != nil {
-				executePlan, err = s.getExecutePlanByBucketID(migrateBucketEvents.CompleteEvents.BucketId.Uint64())
-				// TODO check db should be migrated
-				if err != nil {
-					log.Errorw("bucket migrate schedule received EventCompleteMigrationBucket")
-					continue
-				}
+		for _, unit := range executePlan.gvgUnitMap {
+			if unit.migrateStatus != Migrated {
+				log.Errorw("report task may error, unit should be migrated", "unit", unit)
+			}
+		}
+		s.deleteExecutePlanByBucketID(migrateBucketEvents.CompleteEvents.BucketId.Uint64())
+		executePlan.stopSPSchedule()
+		return err
+	}
+	// 3. process Events
+	if migrateBucketEvents.Events != nil {
+		if s.executePlanIDMap[migrateBucketEvents.Events.BucketId.Uint64()] != nil {
```

Comment:
Did you forget to check if the bucket status is migrating?

Reply:
When a bucket has been migrated, the events of the migration will be listed here, and unexpected repeated migrations will be produce.